proposals/collection-literals.md

```diff
@@ -151,70 +156,72 @@ _Open issue: Relying on a parameter named `capacity` seems brittle. Is there an
+An error is reported for a collection literal if there are multiple `[CollectionBuilder]` attributes on the collection type.
 
-* For example, if an `init void Construct(T[] values)` instance method were added to [`ImmutableArray<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1), then it would be possible for the compiler to emit the following:
+The *create method* must be a `static` method defined on the *builder type*.
+The *create method* must have a single parameter of a type `System.ReadOnlySpan<T0>` passed by value, and a return value of the *collection type* to which the attribute is applied.
 
-    ```c#
-    T[] __storage = /* initialized using predefined rules */
-    ImmutableArray<T> __result = new ImmutableArray<T>();
-    __result.Construct(__storage);
-    ```
+The *builder type* cannot be generic.
+The *builder method* can be a generic *method* and the *arity* of the builder method must match the *arity* of the *collection type*.
 
-    `ImmutableArray<T>` would then take that array directly and use it as its own backing storage.  This would be safe because the compiler (following the requirements around `init`) would ensure that no other location in the code would have access to this temporary array, and thus it would not be possible to mutate it behind the back of the `ImmutableArray<T>` instance.
+Overloads with distinct signatures on the builder type, and overloads on base types, are ignored.
 
-    The above also demonstrates that this approach can work with struct types which do not have a [*parameterless struct constructor*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/parameterless-struct-constructors.md).  In the above, the call to `new ImmutableArray<T>()` is equivalent to `default(ImmutableArray<T>)`, (producing an `ImmutableArray<T>` whose [`IsDefault`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1.isdefault) property is initially true.  However, the `Construct` method can then safely update this to the final non-default state without that intermediate state being visible.
+An error is reported for a collection literal if the `[CollectionBuilder]` attribute on the collection type does not represent an invocable method with the expected signature defined on the *builder type*.
 
-* This formalization is quite beneficial because the only existing mechanism to (safely) create an ImmutableArray with values without copying is both excessively verbose and produces unavoidable garbage:
+If the parameter is `scoped`, the compiler *may* allocate the storage for the `ReadOnlySpan<T0>` on the stack rather than the heap.
```

Comment:
I would rephrase this as "if the parameter does not escape the Create method". Alternatively we could define a concept of "implicitly scoped" and use it here. All I really want to avoid terminology-wise is having to say in several places "if the method does not return ref struct or have a writable ref parameter of ref struct type etc..."

I'm OK with this limitation as it applies to user-defined ref struct types. I somewhat feel that if you are creating a ref struct which is simply wrapping a span, what are you really accomplishing. I don't know whether that's a real use case.

If such ref structs were a real use case, though, my feeling would be that this is unnecessarily limiting. Why shouldn't my ref-struct be able to hold on to stackalloc'd memory just like a Span would. In that case I would prefer a rule that if a collection-expr is converted to ref-struct, its safe-to-escape is the current local scope. It's a simple rule which remains "intact" even through incidental refactorings. If you don't want that behavior, convert the CL to non-ref-struct type first. This is essentially what we suggested for CLs converted directly to Span, I think.